### DISCUSSION TITLE
Is it possible to Toggle Focus?

### DISCUSSION THREAD

**Discussion by timii** [nvim-tree/nvim-tree.lua#1672] [Q&A] [answered]
Hi,

I'm currently using `:NvimTreeFocus` to focus the file explorer when editing a file, but I can't toggle the focus back to the current file without selecting a file from the file explorer. 

My question now is, if it's possible to toggle the focus between the current file and file explorer, without selecting a file in the file explorer, or not? And if it's possible, how can I do it?

I'm sorry if the same question has been answered already, but I couldn't find an answer yet.

If any other information is needed, I will gladly try to provide as much as I can.

Any help is greatly appreciated, thanks.

**Comment by gegoune** [comment, marked as answer]
You can use `wincmd p` within nvim-tree to switch back to previous window. But after quick test I couldn't get it to work (always switching to the next window on the right instead of previous one).

@alex-courtis should that work:
```lua
{ key = 'some-key', action_cb = function() vim.cmd.wincmd 'p' end }
```
? It works fine executed from command line.

**Reply by timii** [reply]
Thank you for your quick answer. I tried using the custom mapping you provided, but I just got an error. After trying around a bit more, it seems to be working after changing `action_cb` to just `action`. My current mapping looks like this:
```lua
{ key = 'some-key', action = function() vim.cmd.wincmd 'p' end }
```

Thanks for your help, and do you have an idea why using `action_cb` doesn't work?

**Reply by alex-courtis** [reply]
I've tried out the above and that results in a failure with a stack: #1675

I have a suspicion that this appeared to work as the custom action was being interpreted as the default, which is open, which would result in the change in focus.

`action` is needed for a mapping; it can be any string you like. Something like this works:

```lua
{ key = 'p', action = 'prev_win', action_cb = function() vim.cmd.wincmd 'p' end }
```

**Reply by timii** [reply]
Interesting, thanks for your response. Does your provided snippet result in a different behavior or any advantages, or does it not matter which one to use?

**Reply by alex-courtis** [reply]
The first one is a side effect that will break once you upgrade to c4ac723

The second will move to the previous window, rather than the window of the file under the cursor.

**Reply by gegoune** [reply]
I ended up with single 'standalone' mapping:
```lua

vim.keymap.set('n', '<leader>f', function()
  if vim.fn.bufname():match 'NvimTree_' then
    vim.cmd.wincmd 'p'
  else
    nvimtree.find_file(true)
  end
end, { desc = 'nvim-tree: toggle' })
```

**Reply by alex-courtis** [reply]
`vim.keymap.set` via `on_attach` is the desired end state and works well today.

The full implementation with all the conveniences is still a work in progress. #1579

Clarified help: https://github.com/nvim-tree/nvim-tree.lua/commit/58055a0397467376b618dede597e458b73c8b4d3

**Reply by rohit-kumar-j** [reply]
@timii, you can try this:

```lua
function toggle_nvimtree()
	if vim.fn.bufname():match 'NvimTree_' then
		vim.cmd.wincmd 'p'
	else
		vim.cmd('NvimTreeFindFile')
	end
end

-- Which-Key
e  = { "<cmd>:lua toggle_nvimtree()<CR>", "Explorer Focus Toggle without closing" },
r  = { "<cmd>:lua require('nvim-tree.api').tree.toggle(false, true)<CR>", "Explorer Peek" },
```
